### PR TITLE
Rework ElimByName

### PR DESCRIPTION
**Pull request by odersky**
Rework ElimByName, so that all transformations are now done in one phase, which 
implements all of the following transformations:

 1. For types of method and class parameters:

    => T     becomes    () ?=> T

 2. For references to cbn-parameters:

    x        becomes    x.apply()

 3. For arguments to cbn parameters

    e        becomes    () ?=> e

 An optimization is applied: If the argument `e` to a cbn parameter is already
 of type `() ?=> T` and is a pure expression, we avoid (2) and (3), i.e. we
 pass `e` directly instead of `() ?=> e.apply()`.

 Note that `() ?=> T`  cannot be written in source since user-defined context functions
 must have at least one parameter. We use the type here as a convenient marker
 of something that will erase to Function0, and where we know that it came from
 a by-name parameter.

 Note also that the transformation applies only to types of parameters, not to other
 occurrences of `ExprType`s. In particular, embedded occurrences in function types
 such as `(=> T) => U` are left as-is here (they are eliminated in erasure).
 Trying to convert these as well would mean traversing all the types, and that
 leads to cyclic reference errors in many cases.  This can cause problems in that
 we might have sometimes a `() ?=> T` where a `=> T` is expected. To compensate,
 there is a new clause in TypeComparer#subArg that declares `() ?=> T` to be a
 subtype of `=> T` for arguments of type applications at any point after this phase
 and up to erasure.

### CODE REVIEW

**Comment by odersky** [comment]
Can someone help figure out why the analyzer plugin test fails? @liufengyun do you have an idea?

**Comment by olhotak** [comment]
In `./sbt-test/sbt-dotty/analyzer-plugin/plugin/Analyzer.scala` line 34:
`override val runsBefore = Set(ElimRepeated.name)`
Probably should be updated to `ProtectedAccessors`.

**Comment by odersky** [comment]
@olhotak That's probably it. Thanks for figuring it out!

File: sbt-test/sbt-dotty/analyzer-plugin/plugin/Analyzer.scala [hunk 1]
```diff
@@ -31,7 +31,7 @@ class InitChecker extends PluginPhase {
   val phaseName = "symbolTreeChecker"
 
   override val runsAfter = Set(SetDefTree.name)
-  override val runsBefore = Set(ElimRepeated.name)
+  override val runsBefore = Set(ProtectedAccessors.name)
```
Comment (liufengyun) [inline sbt-test/sbt-dotty/analyzer-plugin/plugin/Analyzer.scala:34]: `ProtectedAccessors` needs to be imported at line 17.

**Comment by odersky** [comment]
This PR now also contains a rework of RefChecks:

 - Split out forward reference checks and cross version (i.e. experimental/deprecated) checks
   into their own miniphases. These have nothing to do with the core tasks of RefChecks.
 - Move RefChecks earlier in the pipeline, so that it is not affected by ElimByName's questionable
   type manipulations.

RefChecks is one of the last things that we ported from Scala 2 and left largely unchanged. Despite begin a miniphase formally, it was really the typical Scala 2 megaphase that mixed many different things. We could also go further and break out other things. There are still some checks that have nothing to do with overriding or overloading: checkImplicitNotFoundAnnotation and checkUnaryMethods, for instance.

**Review by prolativ — APPROVED** [review]

**Pull request merged**